core/WASM_core.md

```diff
@@ -0,0 +1,352 @@
+}
+
+struct Buckets {
+    buckets: Vec<f64>,
+}
+
+// -------------- Logging --------------
+
+// In Core, all logging is represented as events on trace spans. However, adding a full tracing
+// API would be insanely huge, and way, way too much work on both sides. Additionally, OTel simply
+// does not seem designed for exporting and importing traces in the same process (I can't find any
+// in-memory importers). Even if that worked, just defining the Span struct is significant. All
+// that in mind, I propose only adding a logging interface here.
+
+// This is copied straight out of Core today
+pub struct CoreLog {
+    /// The module within core this message originated from
+    pub target: String,
+    /// Log message
+    pub message: String,
+    /// Time log was generated
+    pub timestamp: SystemTime,
+    /// Message level
+    pub level: Level,
+    /// Arbitrary k/v pairs (span k/vs are collapsed with event k/vs here).
+    pub fields: HashMap<String, serde_json::Value>,
+    /// A list of the outermost to the innermost span names
+    pub span_contexts: Vec<String>,
+}
+
+enum Level {
+    Trace,
+    Debug,
+    Info,
+    Warn,
+    Error,
+}
+
+// The same assumption about a worker-per-module-instance applies here as with metrics.
+fn log(log: CoreLog) {}
+```
+
+#### An aside on serialization
+
+We are presented with an opportunity here to get rid of the ser/de overhead currently present
+as a result of using protos for everything. Since we are introducing additional overhead b/c of
+WASM, we could maybe introduce some savings here. My **completely unsubstantiated** guess is that if
+we stopped the unneeded ser/de we might actually get slightly *faster* under WASM which would be
+amusing.
+
+The options here are Cap'n'Proto or Flatbuffers. Both appear to have support for all the languages
+we care about, except for Flatbuffers not having a Ruby implementation.
+
+Flatbuffer's design is more oriented towards what we're doing, whereas Cap'n'Proto is more concerned
+with sending things over the wire, but is perhaps a bit more ergonomic.
+
+In either case, the primary pain point would be figuring out how to deal with the annoyance of
+having lifetime annotations in the Rust code. Cap'n'Proto has built-in support for creating owned
+versions of the structs which makes that pretty easy, but at that point it's unclear if whatever
+speedup we'd get is worth it compared to Protobuf.
+
+Ultimately, it may makes sense to defer this work until we have measured performance and decided
+it's worth it. Of course, the rub there is adding this support later is a pretty monster change.
+
+What I do think we should take some time to do is make a toy application with the runtime we land on
+(say, with Go as the host), and then just shuffle back and forth some Workflow/Activity tasks using
+proto and using Cap'n/Flatbuffer and measure performance. Unless there's a striking difference, we
+can stick with proto.
+
+# Drawbacks
```

Comment:
Another big one to add/discuss: "fixed memory size". WASM modules are given a fixed-size memory to work with by the WASM runtime (they have pre-defined fixed-size memory blocks and can "import" memory I believe which the runtime can provide, but both are a fixed size IIRC). This number basically has to get set properly from the beginning and we may/should expose it to users, but it's definitely not dynamic. Guess to high and it's unused memory (most WASM runtimes prealloc this, it's not dynamic), guess too low and core will probably panic I figure when it tries to alloc.

Reply:
Rust allows override the `#[global_allocator]`.
I've seen `buddy_alloc` used for that purpose with pre-allocated memory.
But as you've mentioned, the runtime may also provide memory blocks.
AFAICT, externally provided memory can be use for dynamic allocation. See for reference https://developer.mozilla.org/en-US/docs/WebAssembly/JavaScript_interface/Memory.

Reply:
:+1: I was misremembering here based on a JVM limitation I had with `ByteBuffer` (which cannot dynamically grow)